src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -622,6 +606,39 @@ public Task InvokeAsync(object[] parameters)
             }
         }
 
+        private class InvocationHandlerList
+        {
+            private readonly List<InvocationHandler> _invocationHandlers;
+            internal InvocationHandler[] CopiedHandlers;
+
+            internal InvocationHandlerList(InvocationHandler handler)
+            {
+                this._invocationHandlers = new List<InvocationHandler>() { handler};
+                CopiedHandlers = _invocationHandlers.ToArray();
```

Comment:
We should lazy initialize this

Reply:
if we don't do this lazy, we dont have to lock

Reply:
The list isn't generally modified after you start invoking handlers (since generally the `.On` calls all occur at startup, before `StartAsync` is called). It'll be rare to have contention on this lock since the collection isn't usually modified once the `CopiedHandlers` value is requested.

Also, it strikes me that the 90% use-case for this is to have only a single handler. We could probably avoid the allocation of the list altogether by having a field on this list represent the "first" item, and only allocate the list when a second handler is added. That way, in the 90% case, we don't have an extra list hanging around.